File: ingest.py
"""
ingest.py — Load, chunk, embed, and store health knowledge in ChromaDB.
Run once before launching the app: python ingest.py
"""

import os
import sys
# pyrefly: ignore [missing-import]
import chromadb
# pyrefly: ignore [missing-import]
from sentence_transformers import SentenceTransformer
# pyrefly: ignore [missing-import]
from langchain_text_splitters import RecursiveCharacterTextSplitter

# ─── Configuration ────────────────────────────────────────────────────────────
DATA_FILE = os.path.join(os.path.dirname(__file__), "data", "health_knowledge.txt")
CHROMA_PATH = os.path.join(os.path.dirname(__file__), "chroma_db")
COLLECTION_NAME = "health_kb"
EMBEDDING_MODEL = "sentence-transformers/all-MiniLM-L6-v2"
CHUNK_SIZE = 400          # characters per chunk
CHUNK_OVERLAP = 60        # overlapping characters between chunks
# ──────────────────────────────────────────────────────────────────────────────


def load_text(filepath: str) -> str:
    """Load the raw health knowledge text file."""
    if not os.path.exists(filepath):
        print(f"❌  Data file not found: {filepath}")
        sys.exit(1)
    with open(filepath, "r", encoding="utf-8") as f:
        return f.read()


def chunk_text(text: str) -> list[str]:
    """Split text into overlapping chunks for better retrieval coverage."""
    splitter = RecursiveCharacterTextSplitter(
        chunk_size=CHUNK_SIZE,
        chunk_overlap=CHUNK_OVERLAP,
        separators=["\n\n", "\n", ". ", " ", ""],
    )
    chunks = splitter.split_text(text)
    # Filter out very short/empty chunks
    chunks = [c.strip() for c in chunks if len(c.strip()) > 50]
    return chunks


def embed_and_store(chunks: list[str]) -> None:
    """Generate embeddings and store them in ChromaDB."""
    print(f"🔄  Loading embedding model: {EMBEDDING_MODEL}")
    model = SentenceTransformer(EMBEDDING_MODEL)

    print(f"🔄  Connecting to ChromaDB at: {CHROMA_PATH}")
    client = chromadb.PersistentClient(path=CHROMA_PATH)

    # Delete existing collection if it exists (fresh ingest)
    existing = [c.name for c in client.list_collections()]
    if COLLECTION_NAME in existing:
        print(f"ℹ️   Existing collection '{COLLECTION_NAME}' found — recreating for fresh ingest.")
        client.delete_collection(COLLECTION_NAME)

    collection = client.create_collection(
        name=COLLECTION_NAME,
        metadata={"hnsw:space": "cosine"},   # cosine similarity
    )

    print(f"🔄  Generating embeddings for {len(chunks)} chunks…")
    embeddings = model.encode(chunks, show_progress_bar=True, batch_size=32).tolist()

    ids = [f"chunk_{i}" for i in range(len(chunks))]
    metadatas = [{"source": "health_knowledge.txt", "chunk_index": i} for i in range(len(chunks))]

    # Batch upsert
    BATCH = 100
    for start in range(0, len(chunks), BATCH):
        end = min(start + BATCH, len(chunks))
        collection.add(
            ids=ids[start:end],
            documents=chunks[start:end],
            embeddings=embeddings[start:end],
            metadatas=metadatas[start:end],
        )

    print(f"\n✅  Ingestion complete!")
    print(f"    • Total chunks stored : {len(chunks)}")
    print(f"    • Collection name     : {COLLECTION_NAME}")
    print(f"    • ChromaDB path       : {CHROMA_PATH}")


def main():
    print("=" * 55)
    print("  🏥  Health Chatbot — Knowledge Base Ingestion")
    print("=" * 55)

    print(f"\n📂  Loading data from: {DATA_FILE}")
    raw_text = load_text(DATA_FILE)
    print(f"    • File size: {len(raw_text):,} characters")

    print(f"\n✂️   Chunking text (size={CHUNK_SIZE}, overlap={CHUNK_OVERLAP})…")
    chunks = chunk_text(raw_text)
    print(f"    • Total chunks: {len(chunks)}")
    print(f"    • Sample chunk:\n      '{chunks[0][:120]}…'\n")

    embed_and_store(chunks)


if __name__ == "__main__":
    main()

File: server/scripts/ingest.js
/**
 * ingest.js — Load, chunk, embed, and store health knowledge in MongoDB or local JSON.
 * Run once: node scripts/ingest.js
 */
import 'dotenv/config';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import connectDB, { db } from '../config/db.js';
import { generateEmbeddings } from '../services/geminiService.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const DATA_FILE = path.join(__dirname, '..', 'data', 'health_knowledge.txt');
const CHUNK_SIZE = 400;
const CHUNK_OVERLAP = 60;

/**
 * Split text into overlapping chunks.
 */
function chunkText(text) {
  const paragraphs = text.split(/\n\n+/);
  const rawChunks = [];

  let currentChunk = '';
  for (const para of paragraphs) {
    if (currentChunk.length + para.length > CHUNK_SIZE && currentChunk.length > 0) {
      rawChunks.push(currentChunk.trim());
      const overlapText = currentChunk.slice(-CHUNK_OVERLAP);
      currentChunk = overlapText + ' ' + para;
    } else {
      currentChunk += (currentChunk ? '\n\n' : '') + para;
    }
  }
  if (currentChunk.trim().length > 50) {
    rawChunks.push(currentChunk.trim());
  }

  return rawChunks.filter((c) => c.length > 50);
}

async function main() {
  console.log('='.repeat(55));
  console.log('  🏥  Health Chatbot — Knowledge Base Ingestion');
  console.log('='.repeat(55));

  // Initialize DB (MongoDB or fallback)
  await connectDB();

  // Load data
  console.log(`\n📂 Loading data from: ${DATA_FILE}`);
  if (!fs.existsSync(DATA_FILE)) {
    console.error(`❌ Data file not found: ${DATA_FILE}`);
    process.exit(1);
  }
  const rawText = fs.readFileSync(DATA_FILE, 'utf-8');
  console.log(`   • File size: ${rawText.length.toLocaleString()} characters`);

  // Chunk
  console.log(`\n✂️  Chunking text (size=${CHUNK_SIZE}, overlap=${CHUNK_OVERLAP})…`);
  const chunks = chunkText(rawText);
  console.log(`   • Total chunks: ${chunks.length}`);
  console.log(`   • Sample chunk:\n     '${chunks[0].slice(0, 120)}…'\n`);

  // Clear existing chunks
  console.log('🗑️  Clearing existing knowledge base…');
  await db.clearChunks();

  // Generate embeddings in batches
  const BATCH_SIZE = 20;
  console.log(`\n🔄 Generating embeddings for ${chunks.length} chunks (batch size: ${BATCH_SIZE})…`);

  const allEmbeddings = [];
  for (let i = 0; i < chunks.length; i += BATCH_SIZE) {
    const batch = chunks.slice(i, i + BATCH_SIZE);
    process.stdout.write(`   Batch ${Math.floor(i / BATCH_SIZE) + 1}/${Math.ceil(chunks.length / BATCH_SIZE)}… `);
    const embeddings = await generateEmbeddings(batch);
    allEmbeddings.push(...embeddings);
    console.log('✅');

    // Small delay to avoid rate limiting
    if (i + BATCH_SIZE < chunks.length) {
      await new Promise((r) => setTimeout(r, 500));
    }
  }

  // Store in database
  console.log(`\n💾 Storing ${chunks.length} chunks in database…`);
  const docs = chunks.map((text, i) => ({
    text,
    embedding: allEmbeddings[i],
    source: 'health_knowledge.txt',
    chunkIndex: i,
  }));

  await db.addChunks(docs);

  console.log(`\n✅ Ingestion complete!`);
  console.log(`   • Total chunks stored : ${chunks.length}`);
  process.exit(0);
}

main().catch((err) => {
  console.error('❌ Ingestion failed:', err);
  process.exit(1);
});
